### DISCUSSION TITLE
Using middleware inside actions (proposal)

### DISCUSSION THREAD

**Discussion by redbastie** [livewire/livewire#3087] [7. Bugs (archived)]
I can submit a PR for this, but I want to hear some feedback first.

I often find myself using middleware-like behaviour inside of my actions.

For example, I show a button to ALL users, but only AUTHED users can actually use it. This is like a call to action to motivate users to register on the app.

So this code:

    public function follow()
    {
        if (Auth::guest()) {
            return redirect()->route('login');
        }

        $this->user->followers()->attach(Auth::user()->id);
    }

In this action I am allowing users to "follow" other users. If they are not logged in, it sends them to the login route.

Now, it would be much nicer if I could do something like this:

    public function follow()
    {
        $this->middleware('auth');
        $this->user->followers()->attach(Auth::user()->id);
    }

Or even via a property of the component itself:

	public $middleware = [
		'follow' => ['auth'],
	];

Thoughts?

**Comment by oliverbj** [comment]
This would be a very nice feature to have in Livewire! I am currently doing the same as you, where I do the auth checks inside each method, causing a lot of duplication and bloated code.

**Comment by jeffchown** [comment]
For the scenario you’ve described, since Livewire actions are similar to Laravel controller methods, I use policies (https://laravel.com/docs/8.x/authorization#creating-policies).  This way, all my auth code is stored in the policies code which I can call from anywhere via calls like `Gate::authorize('follow');`

**Comment by redbastie** [comment]
> For the scenario you’ve described, since Livewire actions are similar to Laravel controller methods, I use policies (https://laravel.com/docs/8.x/authorization#creating-policies). This way, all my auth code is stored in the policies code which I can call from anywhere via calls like `Gate::authorize('follow');`

This is good. My only issue is basically writing a policy just for checking if the user is currently logged in. Not a huge deal though.

**Comment by cbaconnier** [comment]
It would be beneficial for me. In the application I'm building I have use two guards and two middleware groups (one for each).

Right now it seems to only accept one middleware group for the whole application so I use  `['web', 'auth:web,screen']` 

But allowing components to use middlewares would allow me to do this
 - `['web', 'auth:web']` 
 - `['screen', 'auth:screen']`

**Comment by redbastie** [comment]
> It would be beneficial for me. In the application I'm building I have use two guards and two middleware groups (one for each).
> 
> Right now it seems to only accept one middleware group for the whole application so I use `['web', 'auth:web,screen']`
> 
> But allowing components to use middlewares would allow me to do this
> 
> * `['web', 'auth:web']`
> * `['screen', 'auth:screen']`

You can apply middleware to livewire components via the routes file.

**Comment by cbaconnier** [comment]
> > It would be beneficial for me. In the application I'm building I have use two guards and two middleware groups (one for each).
> > Right now it seems to only accept one middleware group for the whole application so I use `['web', 'auth:web,screen']`
> > But allowing components to use middlewares would allow me to do this
> > 
> > * `['web', 'auth:web']`
> > * `['screen', 'auth:screen']`
> 
> You can apply middleware to livewire components via the routes file.

Yes and no. It works when you hit the URL, but when you poll with Livewire or call an action, it will use `middleware_group` by calling [this route](https://github.com/livewire/livewire/blob/1a1d43ff365301ae99ed7f0ccf02c553d4eeba03/src/LivewireServiceProvider.php#L127)

**Comment by joshhanley** [comment]
@cbaconnier  using middeware_group means you can't have specific middleware per component, that will be for all components.

But since 2.3.9, check out "Persistent Middleware" section on this page https://laravel-livewire.com/docs/2.x/security

Hope this helps!

**Comment by clarkewing** [comment]
> But since 2.3.9, check out "Persistent Middleware" section on this page https://laravel-livewire.com/docs/2.x/security

My main issue with this has to do with testing. I can't pinpoint how to get `Livewire::test(...)` to use middleware.

**Comment by joshhanley** [comment]
@clarkewing I think what you are experiencing is a problem, but this thread is more of a proposal than an issue. Can you open an issue with your problem and details of what you are trying to achieve so it can be investigated?

**Comment by cbaconnier** [comment]
@joshhanley Thanks! But I don't think it helps me. Even If I do register them.
E.g. `auth:screen` won't be re-applied as expected as the sessions table will not be updated but will be when I use `auth:screen` as middleware.

But after re-reading your issue, mine is a bit different indeed. My scenario would be to able to apply middleware to any component or sub-component individually like `protected $middlewares = ['screen', 'auth:screen]`; and an another component `protected $middlewares = ['web', 'auth:web]` as we can do in a controller.
So each time a request would be done, thoses middleware would be called.
